aptos-move/framework/src/natives/aggregator_natives/helpers_v2.rs

```diff
@@ -0,0 +1,51 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_aggregator::aggregator_extension::AggregatorID;
+use move_binary_format::errors::PartialVMResult;
+use move_vm_types::values::{Reference, StructRef, Value};
+
+/// Indices of `value` and `limit` fields in the `Aggregator` Move
+/// struct.
+const VALUE_FIELD_INDEX: usize = 0;
+const LIMIT_FIELD_INDEX: usize = 1;
+
+/// Given a reference to `Aggregator` Move struct returns a field value at `index`.
+pub(crate) fn get_aggregator_field(aggregator: &StructRef, index: usize) -> PartialVMResult<Value> {
+    let field_ref = aggregator.borrow_field(index)?.value_as::<Reference>()?;
+    field_ref.read_ref()
+}
+
+/// Returns ID and a limit of aggrgegator based on a reference to `Aggregator` Move struct.
+pub(crate) fn aggregator_info_u128(
+    aggregator: &StructRef,
+) -> PartialVMResult<(AggregatorID, u128)> {
+    let (value, limit) = get_aggregator_fields_u128(aggregator)?;
+    assert!(
+        value <= u64::MAX as u128,
+        "identifier in aggregator exceeds u64::MAX"
+    );
+    Ok((AggregatorID::ephemeral(value as u64), limit))
+}
+
+/// Returns ID and a limit of aggrgegator based on a reference to `Aggregator` Move struct.
+pub(crate) fn aggregator_info_u64(aggregator: &StructRef) -> PartialVMResult<(AggregatorID, u64)> {
+    let (value, limit) = get_aggregator_fields_u64(aggregator)?;
+    Ok((AggregatorID::ephemeral(value), limit))
+}
+
+/// Given a reference to `Aggregator` Move struct, returns a tuple of its
+/// fields: (`value`, `limit`).
+pub fn get_aggregator_fields_u128(aggregator: &StructRef) -> PartialVMResult<(u128, u128)> {
```

Comment:
This seems identical to `u64` case. Can we avoid it by making generic or using macros?

Reply:
should just

```
  pub fn get_aggregator_fields<T>(aggregator: &StructRef) -> PartialVMResult<(T, T)> {
    let value = get_aggregator_field(aggregator, VALUE_FIELD_INDEX)?.value_as::<T>()?;
    let limit = get_aggregator_field(aggregator, LIMIT_FIELD_INDEX)?.value_as::<T>()?;
    Ok((value, limit))
  }
  
```
work?

what does value_as do if it is given a type different from the one stored? does it convert it/cast it, or fails, or?

Reply:
Implemented your code just now. It gives the following error, so I reverted back to the old code. 
```
the trait `move_vm_types::values::VMValueCast<T>` is not implemented for `move_vm_types::values::Value`
```